### PR TITLE
Adds Boundary Worker

### PR DESCRIPTION
**Pull request by ksatirli**
Hey @workloads/reviewers-nomad,

This PR adds a Pack for Boundary Worker. It handles the full lifecycle (token retrieval, worker registration, worker deregistration) and announces itself correctly with HCP Boundary.

### CODE REVIEW

**Comment by ksatirli** [comment]
@sofixa when you review this: I'm currently working on the `helper` part with @joatmon08; but would love to get your initial thoughts to make sure I'm not missing major things.

File: packs/boundary_worker/README.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+### Application
+
+This section describes Application-specific configuration.
+
+| Name                                     | Description                                                                                                        | Default |
+| ---------------------------------------- | ------------------------------------------------------------------------------------------------------------------ | ------- |
+| app_boundary_helper_output_file_mode     | File Mode of the Output File created by the Boundary Helper binary.                                                | `"0644"` |
+| app_boundary_helper_path                 | Path to the Boundary Helper binary.                                                                                | `"boundary-helper"` |
+| app_cors_allowed_origins                 | Allowed CORS Origins for the Boundary Worker.                                                                      | `["*"]` |
+| app_cors_enabled                         | Toggle to enable CORS support for the Boundary Worker.                                                             | `true` |
```
Comment (sofixa) [inline packs/boundary_worker/README.md:59]: Why is this needed? What's the impact if it's changed?

File: packs/boundary_worker/README.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+| app_disable_mlock                        | Toggle to disable MLock for the Boundary Worker.                                                                   | `true` |
```
Comment (sofixa) [inline packs/boundary_worker/README.md:60]: Same as above, what's the impact? (and I would presume that mlocking wouldn't work with the `exec` task driver without some extra capabilities)

File: packs/boundary_worker/README.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+| app_enable_hcp_boundary_support          | Toggle to enable HCP Boundary Support (and forego self-hosted Boundary Enterprise Cluster registration workflows.  | `true` |
```
Comment (sofixa) [inline packs/boundary_worker/README.md:61]: Above in the README it says an HCP cluster id is needed, but this seems to imply it would work for self-hosted Boundary too?

File: packs/boundary_worker/README.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+| app_initial_upstreams                    | Initial Upstreams for the Boundary Worker.                                                                         | n/a |
```
Comment (sofixa) [inline packs/boundary_worker/README.md:62]: Might be worth throwing a link to the Boundary doc about what a Boundary Upstream is

File: packs/boundary_worker/README.md [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+| app_tls_client_ca_file                   | PEM-encoded Certificate Authority File used for checking the authenticity of tthe client for the Boundary Worker.  | n/a |
+| app_tls_disable                          | Toggle to disable TLS support for the Boundary Worker.                                                             | `true` |
+| app_tls_key_file                         | Specifies the path to the private key for the certificate for the Boundary Worker.                                 | n/a |
+| app_tls_max_version                      | Specifies the maximum supported TLS version for the Boundary Worker.                                               | `"tls13"` |
+| app_tls_min_version                      | Specifies the minimum supported TLS version for the Boundary Worker.                                               | `"tls12"` |
+| app_tls_prefer_server_cipher_suites      | Toggle to enable preference for Server's ciphersuites over Client's ciphersuites for the Boundary Worker.          | `false` |
```
Comment (sofixa) [inline packs/boundary_worker/README.md:70]: This probably only works with TLS non 1.3, which should be the default, so might be worth noting this.
https://go.dev/blog/tls-cipher-suites

File: packs/boundary_worker/README.md [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+| app_tls_cert_file                        | Specifies the path to the certificate for the Boundary Worker.                                                     | n/a |
+| app_tls_cipher_suites                    | Overridden List of supported ciphersuites for the Boundary Worker.                                                 | n/a |
+| app_tls_client_ca_file                   | PEM-encoded Certificate Authority File used for checking the authenticity of tthe client for the Boundary Worker.  | n/a |
+| app_tls_disable                          | Toggle to disable TLS support for the Boundary Worker.                                                             | `true` |
+| app_tls_key_file                         | Specifies the path to the private key for the certificate for the Boundary Worker.                                 | n/a |
+| app_tls_max_version                      | Specifies the maximum supported TLS version for the Boundary Worker.                                               | `"tls13"` |
+| app_tls_min_version                      | Specifies the minimum supported TLS version for the Boundary Worker.                                               | `"tls12"` |
```
Comment (sofixa) [inline packs/boundary_worker/README.md:69]: TLS 1.3 should be the minimum too

File: packs/boundary_worker/README.md [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+| app_tls_prefer_server_cipher_suites      | Toggle to enable preference for Server's ciphersuites over Client's ciphersuites for the Boundary Worker.          | `false` |
+| app_tls_require_and_verify_client_cert   | Toggle to enable client authentication for the listener for the Boundary Worker.                                   | `false` |
+| app_worker_description                   | Description for the Boundary Worker.                                                                               | `"Nomad-managed Boundary Worker."` |
```
Comment (sofixa) [inline packs/boundary_worker/README.md:72]: What's the purpose of the description? Does it get propagated to Boundary?

File: packs/boundary_worker/README.md [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+### Nomad
+
+This section describes Nomad-specific configuration.
+
+| Name                              | Description                                  | Default |
+| --------------------------------- | -------------------------------------------- | ------- |
+| nomad_group_count                 | Count of Deployments for the Group.          | `1` |
+| nomad_group_ephemeral_disk        | Ephemeral Disk Configuration for the Group.  | `{"migrate":true,"size":128,"sticky":true}` |
+| nomad_group_name                  | Name for the Group.                          | `"boundary_worker"` |
+| nomad_group_network_mode          | Network Mode for the Group.                  | `"host"` |
```
Comment (sofixa) [inline packs/boundary_worker/README.md:84]: Might be worth mentioning that without TLS and with network_mode=host, anything on the host would be able to MITM/eavesdrop on the communications between the worker and the control plane, or making the defaults such as that it's hard (maybe an auto generated cert?)

File: packs/boundary_worker/README.md [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+| nomad_group_ports                 | Port Configuration for the Group.            | `{"ops":{"check_interval":"30s","check_timeout":"15s","host_network":null,"name":"boundary_worker_ops","path":"/health","port":9203,"protocol":"http","type":"http"},"proxy":{"check_interval":"30s","check_timeout":"15s","host_network":null,"name":"boundary_worker_proxy","path":null,"port":9202,"protocol":"tcp","type":"tcp"}}` |
```
Comment (sofixa) [inline packs/boundary_worker/README.md:85]:
```suggestion
| nomad_group_ports                 | Port and Healthcheck Configuration for the Group.            | `{"ops":{"check_interval":"30s","check_timeout":"15s","host_network":null,"name":"boundary_worker_ops","path":"/health","port":9203,"protocol":"http","type":"http"},"proxy":{"check_interval":"30s","check_timeout":"15s","host_network":null,"name":"boundary_worker_proxy","path":null,"port":9202,"protocol":"tcp","type":"tcp"}}` |
```

File: packs/boundary_worker/README.md [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+
+> **Note**
+>
+> The outputs are only rendered if `nomad_pack_verbose_output` is set to `true`.
+
+## Notes
+
+### Resource Sizing
+
+For hardware sizing of the underlying Nomad Agents that are designed to run the Boundary Worker in this Nomad Pack, see the [HCP Boundary Self-managed Worker sizing guidelines](https://developer.hashicorp.com/hcp/docs/boundary/self-managed-workers/size-self-managed-workers#sizing-guidelines-for-self-managed-workers).
```
Comment (sofixa) [inline packs/boundary_worker/README.md:114]: Aren't the default resource limits above too small compared to the sizing guidelines?

File: packs/boundary_worker/templates/pack.nomad.tpl [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+      # see https://developer.hashicorp.com/nomad/docs/concepts/workload-identity
+      identity {
+        env = true
+      }
+
+      [[ template "environment_variables" . ]]
+
+      # see https://developer.hashicorp.com/nomad/docs/job-specification/template
+      template {
+        change_mode = "restart"
```
Comment (sofixa) [inline packs/boundary_worker/templates/pack.nomad.tpl:134]: Some config changes can be handled by a reload, but I don't know which ones and if they're enough :D

File: packs/boundary_worker/templates/pack.nomad.tpl [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+
+      # see https://developer.hashicorp.com/nomad/docs/job-specification/volume_mount
+      [[/* iterate over `var.volumes` to create Volume Mounts */]]
+      [[- range $index, $mount := var "nomad_group_volumes" . ]]
+      volume_mount {
+        volume      = "[[ $mount.name ]]"
+        destination = "[[ $mount.destination ]]"
+        read_only   = [[ $mount.read_only ]]
+      }
+      [[ end ]]
```
Comment (sofixa) [inline packs/boundary_worker/templates/pack.nomad.tpl:152]: Is there anything that might be worth preserving or can the volume config be removed entirely?

File: packs/boundary_worker/templates/pack.nomad.tpl [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+
+      # see https://developer.hashicorp.com/nomad/docs/job-specification/resources
+      resources {
+        [[- $resources := var "nomad_task_resources" . ]]
+        cpu        = [[ $resources.cpu ]]
+        cores      = [[ $resources.cores | default "null" ]]
+        memory     = [[ $resources.memory ]]
+
+        # TODO: add support for memory oversubscription
+        # memory_max = [[ $resources.memory_max ]]
```
Comment (sofixa) [inline packs/boundary_worker/templates/pack.nomad.tpl:162]: This is declared in the variables.hcl

File: packs/boundary_worker/README.md (hunk 2 above)
Comment (sofixa) [reply packs/boundary_worker/README.md:60]: And why is it true by default?

File: packs/boundary_worker/variables.hcl [hunk 14]
```diff
@@ -0,0 +1,347 @@
...
+variable "app_tls_key_file" {
+  type        = string
+  description = "Specifies the path to the private key for the certificate for the Boundary Worker."
+  default     = ""
+}
+
+variable "app_tls_min_version" {
+  type        = string
+  description = "Specifies the minimum supported TLS version for the Boundary Worker."
+  default     = "tls12"
```
Comment (sofixa) [inline packs/boundary_worker/variables.hcl:85]:
```suggestion
  default     = "tls13"
```

TLS 1.2 is insecure

File: packs/boundary_worker/variables.hcl [hunk 15]
```diff
@@ -0,0 +1,347 @@
...
+
+variable "app_tls_max_version" {
+  type        = string
+  description = "Specifies the maximum supported TLS version for the Boundary Worker."
+  default     = "tls13"
+}
+
+variable "app_tls_cipher_suites" {
+  type        = string
+  description = "Overridden List of supported ciphersuites for the Boundary Worker."
```
Comment (sofixa) [inline packs/boundary_worker/variables.hcl:96]:
```suggestion
  description = "Overridden List of supported ciphersuites for the Boundary Worker (only applies for TLS1.2 and older, but not on TLS1.3 which is the default)."
```

File: packs/boundary_worker/README.md (hunk 1 above)
Comment (ksatirli) [reply packs/boundary_worker/README.md:59]: Great catch! 

This makes more sense in a non-HCP setting, so I'll set sane defaults for HCP

File: packs/boundary_worker/README.md (hunk 2 above)
Comment (ksatirli) [reply packs/boundary_worker/README.md]: A remnant from building it for non-HCP and speed of development. Updated and removed! ❤️

File: packs/boundary_worker/README.md (hunk 3 above)
Comment (ksatirli) [reply packs/boundary_worker/README.md:61]: We technically - from a functional perspective - support both HCP and self-hosted, but right now focus on HCP-first.

I've updated the README to make it more clear

File: packs/boundary_worker/templates/pack.nomad.tpl (hunk 13 above)
Comment (ksatirli) [reply packs/boundary_worker/templates/pack.nomad.tpl:162]: It is - as is the case for _all_ Packs. It will get updated as soon as all Packs move to a more modular structure.

File: packs/boundary_worker/templates/pack.nomad.tpl (hunk 11 above)
Comment (ksatirli) [reply packs/boundary_worker/templates/pack.nomad.tpl:134]: "restart" is the preferred option here. From the [Worker](https://developer.hashicorp.com/boundary/docs/configuration/worker#signals) docs:

> The SIGHUP signal causes a worker to reload its configuration file to pick up any updates for the `initial_upstreams` and `tags` values. Any other updated values are ignored.

For the primary use-case of this Pack, `initial_upstreams` isn't as important, and `tags` would primarily change if the rendered-and-run Pack would be re-allocated.

File: packs/boundary_worker/variables.hcl (hunk 14 above)
Comment (ksatirli) [reply packs/boundary_worker/variables.hcl:85]: In the [docs](https://developer.hashicorp.com/boundary/docs/configuration/listener/tcp#tls_min_version), we use 1.2 as the default, and offer 1.3.

Reading up more on this, I concur with you that the default _should_ be `tls13`, just to have the most secure default.

File: packs/boundary_worker/variables.hcl (hunk 15 above)
Comment (ksatirli) [reply packs/boundary_worker/variables.hcl:96]: I changed this to something else, but kept it functionally intact

File: packs/boundary_worker/README.md (hunk 4 above)
Comment (ksatirli) [reply packs/boundary_worker/README.md:62]: I cleared up the description, thank you!

File: packs/boundary_worker/templates/pack.nomad.tpl (hunk 12 above)
Comment (ksatirli) [reply packs/boundary_worker/templates/pack.nomad.tpl:152]: I'll take care of this in #23

File: packs/boundary_worker/README.md (hunk 7 above)
Comment (ksatirli) [reply packs/boundary_worker/README.md:72]: It shows up in the Boundary Controller UI as the description of the controller:

Screenshot 2024-01-25 at 11 22 04

File: packs/boundary_worker/README.md (hunk 9 above)
Comment (ksatirli) [reply packs/boundary_worker/README.md:85]: Added!

File: packs/boundary_worker/README.md (hunk 6 above)
Comment (ksatirli) [reply packs/boundary_worker/README.md:69]: Done!

File: packs/boundary_worker/README.md (hunk 2 above)
Comment (ksatirli) [reply packs/boundary_worker/README.md]: I develop this on macOS, where `mlock` triggers a warning, as `boundary` isn't running inside a container, but on the host itself.

I've updated the docs to be more clear - thank you for the spot!

File: packs/boundary_worker/README.md (hunk 10 above)
Comment (ksatirli) [reply packs/boundary_worker/README.md:114]: The sizing guidelines definitely call for more memory and CPU than specified here. 

I think as defaults, these values are fine, but should definitely be tweaked by specifying a more appropriate value when running the Pack.

AWS M5 instances are 8GB of memory, which this Pack does _not_ need, and setting it as the default would make it harder to test it.

File: packs/boundary_worker/README.md (hunk 8 above)
Comment (ksatirli) [reply packs/boundary_worker/README.md:84]: This is a good call out, I'll think about this more in #24
